Update Apache Commons Text to 1.10.0

ParaskP7 opened the pull request:
After an internal discussion (see `peaMlT-2f-p2` and `peaMlT-2f-p2#comment-80`, this PR upgrades Apache's `Commons Text` library to [1.10.0](https://commons.apache.org/proper/commons-text/changes-report.html#a1.10.0) in order fix the possible `StringSubstitutor` related vulnerability in the library.

For more details see: https://commons.apache.org/proper/commons-text/changes-report.html#a1.10.0

And more specifically see release notes entry below:

```
Make default string lookups configurable via system property. Remove
dns, url, and script lookups from defaults. If these lookups are
required for use in StringSubstitutor.createInterpolator(), they must be
enabled via system property. See StringLookupFactory for details.
```

-----

As part of this `commonsText = '1.10.0'` update the below changes were also applied in order to try and strip down as of the Apache Commons dependency as possible:

Remove `Apache Commons` from `Plugins:WooCommerce` module:

1. [Remove unnecessary org apache commons lang dependency for woo.](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2573/commits/ac2e1a4962c0022926d638fe63a725b2ecdc4fcf)

Remove `Apache Commons` from `Example` module:

1. [Replace unnecessary commons lang date utils for example.](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2573/commits/1c082d95ae56be3c4b493093c7dd17349395cc5f)
2. [Replace commons lang array utils contains with array utils.](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2573/commits/292be05c04f28b8fc41e4a5005246b7a6d1a58e9)
3. [Remove unnecessary commons lang array utils contains.](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2573/commits/5e95cd079e1ba1b84db25edd11cf3102d9f6b6cd)
4. [Replace commons lang random utils strings with custom one.](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2573/commits/5e3b9ed85c6dcb5913354ffc997e3e7b6fa3b9b8)
5. [Remove unnecessary commons lang random string utils.](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2573/commits/fafc3e86c156f572e706968dbd7ec65a5d569fdf)

Refactoring List:

1. [Optimize string utils import for site utils.](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2573/commits/a4b468acb20a066f4812e98c79a6f2f35a6ad078)

-----

With this change the only Apache's `Commons Lang` and `Commons Text` related utilities that this project is using are the below:

1. `Commons Lang` -> `org.apache.commons.lang3.StringUtils`
    1. `StringUtils.split(...)` -> See `SiteUtils.java` class.
    2. `StringUtils.equals(...)` -> See `MediaUploadModel.java` class.
2. `Commons Text` -> `org.apache.commons.text.StringEscapeUtils`
    1. `StringEscapeUtils.unescapeHtml4(...)` -> 17 usages in project.
3. (⚠️)* `Commons Lang` -> `org.apache.commons.lang3.StringEscapeUtils`
    1. Deprecated `StringEscapeUtils.escapeXml(...)` -> See `XmlrpcUploadRequestBody`

-----

A quick diff of the previous usage of `commons-lang3-3.8.1` vs. the current usage of `commons-lang3-3.12.0` showed that only the `StringUtils.equals(...)` got updated, see `Step-wise comparison` below:

`commons-lang3-3.8.1`
```
    public static boolean equals(final CharSequence cs1, final CharSequence cs2) {
        ...
        return CharSequenceUtils.regionMatches(cs1, false, 0, cs2, 0, cs1.length());
    }
```

`commons-lang3-3.12.0`
```
    public static boolean equals(final CharSequence cs1, final CharSequence cs2) {
        ...
        // Step-wise comparison
        final int length = cs1.length();
        for (int i = 0; i < length; i++) {
            if (cs1.charAt(i) != cs2.charAt(i)) {
                return false;
            }
        }
        return true;
    }
```

-----

A quick diff of the previous usage of `commons-text-1.1` vs. the current usage of `commons-text-1.10.0` showed that only the inner to `translate(...)` method of `StringEscapeUtils.unescapeHtml4(...)` got slightly updated, see `RuntimeException` vs. `UncheckedIOException` below:

`commons-text-1.1`
```
    public final String translate(final CharSequence input) {
        ...
        } catch (final IOException ioe) {
            // this should never ever happen while writing to a StringWriter
            throw new RuntimeException(ioe);
        }
    }
```

`commons-text-1.10.0`
```
    public final String translate(final CharSequence input) {
        ...
        } catch (final IOException ioe) {
            // this should never ever happen while writing to a StringWriter
            throw new UncheckedIOException(ioe);
        }
    }
```

-----

(*) Warning (⚠️)

About the warning above, note the fact that this project uses two similar imports for `StringEscapeUtils`:
- org.apache.commons.text.StringEscapeUtils ✅ (see 17 usages)
- org.apache.commons.lang3.StringEscapeUtils ⚠️ (see `XmlrpcUploadRequestBody`)

However, note that this `lang3` import is necessary, as the `text` import does not expose the `StringEscapeUtils.escapeXml(...)` api, instead it only exposed the `StringEscapeUtils.escapeXml10(...)` and `StringEscapeUtils.escapeXml11(...)` apis. As such, and until `XmlrpcUploadRequestBody` to either using `xml10` or `xml11`, it is necessary for this class to use the `lang3` import instead of the `text` import. 🤔

In addition to the above, note the `TODO` comment on `XmlrpcUploadRequestBody`, close to the `mPrependString` field and its usage of `StringEscapeUtils.escapeXml(...)` (see below). As such, this means that a proper future solution on that would be to completely remove this custom `String.format(...)` logic and replace that with `XMLRPCSerializer`, that is, instead of going for updating `xml` with either `xml10` or `xml11`. 🔍

`// TODO: we should use the XMLRPCSerializer instead of doing this`

-----

To test:

1. First, quickly smoke test the `FluxC Example` app and see if it works as expected.
2. Then, try and be more specific by taking a look at the code changes, associate those to some screens and make sure everything is working as expected (such an example could be `WCOrderListActivity`).
3. Finally, quickly smoke test, either the `WordPress` or `Woo Commerce` apps, with this version of `FluxC`, and see if it works as expected.

-----

Finally, as a regression note, note the fact that the `MediaUploadModel.equals(...)` method might be producing different results before and after the change, that is, since the `StringUtils.equals(...)` implementation was slightly changed.

ParaskP7 committed:
Deps: Remove unnecessary org apache commons lang dependency for woo

This dependency was added as part of the
445601941d74d1430287e56b76596244f567e469 commit on the
'plugins:woocommerce' module, most probably in order to be later used as
part of the d0b92ceab8aee19de4e2148d40385d49dad20ade commit, but this
time, on the 'example' module on 'WCOrderListActivity' class in order to
import 'org.apache.commons.lang3.time.DateUtils'.

However, the 'example' module already has such a
'org.apache.commons:commons-lang3' dependency declaration. As such, this
similar dependency declaration on the 'plugins:woocommerce' module is
totally unused and can be safely removed.

ParaskP7 committed:
Analysis: Replace unnecessary commons lang date utils for example

The 'DateUtils.addMonth/Weeks/Days(...)' and 'DateUtils.isSameDay(...)'
can be safely replaced with a Kotlin equivalent solution.

As such, this commit also removes the unnecessary org apache commons
lang main dependency for the 'example' module.

FYI: The underlying code was copied from the Apache Commons Lang3
'DateUtils' utility and then converted to Kotlin without any extra
modification whatsoever.

PS: The same change was applied on 'WPAndroid' (see PR/commit below).
As such, and after these changes are confirmed by manual testing, it
could be possible to move this 'DateUtils' into the
'WordPress-Utils-Android' repo and made this as the universal source of
truth, just like it already happens for 'DateTimeUtils' (see
'org.wordpress.android.util.DateTimeUtils' import).

WPAndroid PR: https://github.com/wordpress-mobile/WordPress-Android/
pull/17501
WPAndroid Commit: https://github.com/wordpress-mobile/WordPress-Android/
pull/17501/commits/fcd2cae96c8b55aecde25d2359ac5bd1642b3fcb

ParaskP7 committed:
Analysis: Replace commons lang array utils contains with array utils

This Kotlin related 'ArraysUtils' utility file was created so as to then
use Kotlin's language and replace the Apache Commons Lang3 'ArrayUtils'
utility and thus depend even less on the Apache Commons library going
forward.

ParaskP7 committed:
Analysis: Remove unnecessary commons lang array utils contains

The 'ArrayUtils.contains(...)' can be safely replaced with a Kotlin
equivalent solution.

FYI: The tests within the 'ArrayUtilsTest' test suite verify correctness
of this change.

PS: Part of the same change was applied on 'WPAndroid' (see PR/commit
below). As such, and after these changes are confirmed by manual
testing, it could be possible to move this 'ArrayUtils' into the
'WordPress-Utils-Android' repo and made this as the universal source of
truth, just like it already happens for other such utilities.

WPAndroid PR: https://github.com/wordpress-mobile/WordPress-Android/
pull/17501
WPAndroid Commit: https://github.com/wordpress-mobile/WordPress-Android/
pull/17501/commits/762fb23046fdde78b92bd90687a56bf5736d915b

ParaskP7 committed:
Analysis: Replace commons lang random utils strings with custom one

This Kotlin related 'RandomStringUtils' utility file was created so as
to then use Kotlin's language and replace the Apache Commons Lang3
'RandomStringUtils' utility and thus depend even less on the Apache
Commons library going forward.

ParaskP7 committed:
Analysis: Remove unnecessary commons lang random string utils

The 'RandomStringUtils.randomAlphanumeric/Alphabetic/Number(...)' can be
safely replaced with a Kotlin equivalent solution.

As such, this commit also removes the unnecessary org apache commons
lang Android test dependency for the 'example' module.

FYI: Since this is a random string utility, it is difficult to unit
test that. However, I did manually unit test those for correctness,
compared those to the Apache Commons Lang3 'RandomStringUtils' and
verified that the results are the expected ones. Also, the connected
tests can verify correctness of this change as well.

ParaskP7 committed:
Refactor: Optimize string utils import for site utils

ParaskP7 committed:
Deps: Update apache commons text to 1.10.0

This mainly fixes the possible 'StringSubstitutor' related vulnerability
in Apache's 'Commons Text' library.

For more details see: https://commons.apache.org/proper/commons-text/
changes-report.html#a1.10.0

And more specifically see release notes entry below:

"Make default string lookups configurable via system property. Remove
dns, url, and script lookups from defaults. If these lookups are
required for use in StringSubstitutor.createInterpolator(), they must be
enabled via system property. See StringLookupFactory for details."

oguzkocer commented on example/src/main/java/org/wordpress/android/fluxc/example/utils/DateUtils.kt:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    private fun isSameDay(calendar1: Calendar?, calendar2: Calendar?): Boolean {
+        require(!(calendar1 == null || calendar2 == null)) { "The calendar must not be null" }
```
Using a nullable parameter just to throw an exception for `null` values doesn't seem right to me. We should be able to at least update the private function, and most likely the public ones without too much trouble.

Note that I wanted to bring this up in the [WPAndroid PR](https://github.com/wordpress-mobile/WordPress-Android/pull/17501) as well, but since it was auto-converted, I thought maybe it wasn't worth it. I am still not 100% sure if this should or shouldn't be part of this PR, but I am more inclined to addressing this in both [WPAndroid PR](https://github.com/wordpress-mobile/WordPress-Android/pull/17501) as well as this one. What do you think @ParaskP7?

ParaskP7 replied:
Agreed @oguzkocer ! 💯 

This is now done: 951399cc69b6a9e2a0c18e91c61e7fe4abae8d9c

oguzkocer commented on example/src/main/java/org/wordpress/android/fluxc/example/utils/RandomStringUtils.kt:
```diff
@@ -0,0 +1,44 @@
+        )
+    }
+
+    private fun randomString(count: Int, letters: Boolean, numbers: Boolean): String {
```
This is a personal preference, but I really like having default values for `letters` & `numbers`, so only the specific parameters would need to be configured.

```suggestion
    private fun randomString(count: Int, letters: Boolean = true, numbers: Boolean = true): String {
```

ParaskP7 replied:
I was thinking about that as well, but I ended up not using default values for `letters` & `numbers` as I liked the individual functions being explicit about that, it seemed more readable to me, that is, in terms of making sense of what is happening without needing to look at the private `randomString(...)` function.

This is now done: b1a06eaacf0ee09280306f0a219d698cc34e744b

Let me know how you like your suggested alternative @oguzkocer ! 🙏

oguzkocer reviewed (approved):
@ParaskP7 I left one comment for your consideration in `DateUtils`, but otherwise it looks good to me.

ParaskP7 committed:
Refactor: Make is same day private function accept non-nullable calendar

Since this is a private function and cannot be accessed publicly,
making it accept only non-nullable calendar instances is preferred so
that to avoid unnecessary boilerplate code.

ParaskP7 committed:
Refactor: Add default values to letters and number for random string

ParaskP7 commented:
👋 @oguzkocer !

> @ParaskP7 I left one comment for your consideration in DateUtils, but otherwise it looks good to me.

Great, thank you so much for reviewing and testing the changes, you rock! 🙇 ❤️ 🚀 

PS: I've applied some additional changes based on your comments, let me know what you think of them! 🙏

oguzkocer commented:
Thanks for the latest changes @ParaskP7! I think this is all ready to merge.

ParaskP7 commented:
> Thanks for the latest changes @ParaskP7! I think this is all ready to merge.

Awesome, thanks @oguzkocer ! 🙇
